src/CloudNative.CloudEvents.Amqp/AmqpClientExtensions.cs

```diff
@@ -94,5 +95,96 @@ private static bool HasCloudEventsContentType(Message message, out string conten
+                    bodySection = SerializeData(cloudEvent.Data);
+                    properties = new Properties { ContentType = cloudEvent.DataContentType };
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(contentMode), $"Unsupported content mode: {contentMode}");
+            }
+            return new Message
+            {
+                ApplicationProperties = applicationProperties,
+                BodySection = bodySection,
+                Properties = properties
+            };
+        }
+
+        private static ApplicationProperties MapHeaders(CloudEvent cloudEvent)
+        {
+            var applicationProperties = new ApplicationProperties();
+            var properties = applicationProperties.Map;
+            properties.Add(SpecVersionAmqpHeader, cloudEvent.SpecVersion.VersionId);
+
+            foreach (var pair in cloudEvent.GetPopulatedAttributes())
+            {
+                var attribute = pair.Key;
+
+                // The content type is specified elsewhere.
+                if (attribute == cloudEvent.SpecVersion.DataContentTypeAttribute)
+                {
+                    continue;
+                }
+
+                string propKey = AmqpHeaderPrefix + attribute.Name;
+
+                // TODO: Check that AMQP can handle byte[], bool and int values
+                object propValue = pair.Value switch
+                {
+                    Uri uri => uri.ToString(),
+                    // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
+                    // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
+                    DateTimeOffset dto => dto.UtcDateTime,
+                    _ => pair.Value
+                };
+                properties.Add(propKey, propValue);
+            }
+            return applicationProperties;
+        }
+
+        /// <summary>
+        /// Convert data into a suitable format for inclusion in an Avro record.
+        /// TODO: Asynchronous version?
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        private static RestrictedDescribed SerializeData(object data)
+        {
+            switch (data)
+            {
+                case null:
+                    return null;
+                case byte[] bytes:
+                    return new Data { Binary = bytes };
+                case MemoryStream memoryStream:
+                    // Note: this will return the whole stream, regardless of position...
+                    return new Data { Binary = memoryStream.ToArray() };
+                case Stream stream:
+                    var buffer = new MemoryStream();
+                    stream.CopyTo(buffer);
+                    return new Data { Binary = buffer.ToArray() };
+                case string text:
+                    return new AmqpValue { Value = text };
```

Comment:
The spec says the data should go in the Data section (as opposed to the value section which is what AmqpValue maps to) - https://github.com/cloudevents/spec/blob/v1.0/amqp-protocol-binding.md#13-content-modes

Reply:
Well spotted. Filed https://github.com/cloudevents/sdk-csharp/issues/104 for this. (This PR doesn't change the behavior, just moves it. We'll want more tests in general...)